File: java/src/Leetcode142.java
import java.util.*;
import utils.ListNode;

public class Leetcode142 {
    public ListNode detectCycle(ListNode head){
        Set<ListNode> set = new HashSet<>();
        while(head != null){
            if(set.contains(head))
                return head;
            else{
                set.add(head);
                head = head.next;
            }
        }

        return null;
    }
}

File: python/Leetcode142.py
from ListNode import ListNode
from typing import Optional

class Leetcode142:
    def detectCycle(self, head:Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return None
        visited = set()

        while head is not None:
            if head in visited:
                return head
            else:
                visited.add(head)
                head = head.next
        return None